### PR TITLE
✨ Webserver: Edit templates

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR allows to edit a template provided the user has *write* rights to it.

@odeimaiz will adjust the frontend part either here or in a later PR (lock on template + UI)

image

## How to test

- Create your own template
- Publish it
- go to template tab
- click on ```...```
- click on ```Edit```
- --> this opens the template and allows to modify it.

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/projects/projects_handlers.py [hunk 1]
```diff
@@ -48,13 +52,27 @@ async def open_project(request: web.Request) -> web.Response:
...
+        if project_type is ProjectType.TEMPLATE and user_role < UserRole.USER:
+            # only USERS/TESTERS can do that
+            raise web.HTTPForbidden(reason="Wrong user role to open/edit a template")
+
         project = await projects_api.get_project_for_user(
             request.app,
             project_uuid=f"{path_params.project_id}",
             user_id=req_ctx.user_id,
-            include_templates=False,
+            include_templates=bool(project_type is ProjectType.TEMPLATE),
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:69]: does the is operator not return True or False?

File: services/web/server/tests/unit/with_dbs/02/test_projects_handlers__open_close.py [hunk 2]
```diff
@@ -589,6 +686,73 @@ async def test_close_project(
...
+#                 service_uuid=service["service_uuid"],
+#                 save_state=True,
+#             )
+#             for service in fake_dynamic_services
+#         ]
+#         mocked_director_v2_api[
+#             "director_v2_core_dynamic_services.stop_dynamic_service"
+#         ].assert_has_calls(calls)
+
+#         # should not be callsed request_retrieve_dyn_service
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/02/test_projects_handlers__open_close.py:753]: Did you forget to uncomment this? what happened here?

Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/02/test_projects_handlers__open_close.py:753]: good one.

**Review by odeimaiz — APPROVED** [review]
Merci!

**Review by elisabettai — APPROVED** [review]
Very very cool! 

Any chance we can re-enable the feature to show the link to share a template? 
https://github.com/ITISFoundation/osparc-issues/issues/806

Apologizes if that's out-of-scope from this PR. 😊

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py [hunk 3]
```diff
@@ -463,10 +465,10 @@ async def _get_project(
             SELECT *
             FROM projects
             WHERE
-            {"" if include_templates else "projects.type != 'TEMPLATE' AND"}
+            {"projects.type = 'TEMPLATE' AND" if only_templates else ""}
             uuid = '{project_uuid}'
             AND (jsonb_exists_any(projects.access_rights, {_assemble_array_groups(user_groups)})
-            OR prj_owner = {user_id})
+            OR prj_owner = {user_id} {"OR published='true'" if only_published else ""})
```
Comment (pcrespov) [inline services/web/server/src/simcore_service_webserver/projects/projects_db.py:471]: uff, this text expression is getting pretty. I guess at some point we should create the query with sqlalchemy query language?

**Review by pcrespov — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/studies_dispatcher/_studies_access.py [hunk 4]
```diff
@@ -52,8 +53,8 @@ async def get_public_project(app: web.Application, project_uuid: str):
     """
     from ..projects.projects_db import APP_PROJECT_DBAPI
 
-    db = app[APP_PROJECT_DBAPI]
-    prj = await db.get_template_project(project_uuid, only_published=True)
+    db: ProjectDBAPI = app[APP_PROJECT_DBAPI]
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/studies_dispatcher/_studies_access.py:56]: did not know of that one. thanks!

**Comment by sanderegg** [comment]
> Very very cool!
> 
> Any chance we can re-enable the feature to show the link to share a template? [ITISFoundation/osparc-issues#806](https://github.com/ITISFoundation/osparc-issues/issues/806)
> 
> Apologizes if that's out-of-scope from this PR. blush
@elisabettai 
I am not sure about that one. is this linked to backend or frontend @odeimaiz ?

**Comment by odeimaiz** [comment]
> > I am not sure about that one. is this linked to backend or frontend @odeimaiz ?

I think it's related to the published flag

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py [hunk 5]
```diff
@@ -463,10 +465,10 @@ async def _get_project(
             SELECT *
             FROM projects
             WHERE
-            {"" if include_templates else "projects.type != 'TEMPLATE' AND"}
+            {"projects.type = 'TEMPLATE' AND" if only_templates else ""}
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_db.py:468]: done

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py (hunk 3 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_db.py:471]: it's not scarry, that's just SQL. but I used chatgpt to convert it now to sqlalchemy. let's see if that works well

**Pull request merged**